sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java

```diff
@@ -0,0 +1,318 @@
+                return RequestBody.create(byteString1, okhttp3.MediaType.parse(contentType));
+            }
+        });
+    }
+
+    /**
+     * Aggregate Flux of java.nio.ByteBuffer to single okio.ByteString.
+     *
+     * Pooled okio.Buffer type is used to buffer emitted ByteBuffer instances.
+     * Content of each ByteBuffer will be written (i.e copied) to the internal okio.Buffer slots.
+     * Once the stream terminates, the contents of all slots get copied to one single byte array
+     * and okio.ByteString will be created referring this byte array.
+     * Finally the initial okio.Buffer will be returned to the pool.
+     *
+     * @param bbFlux the Flux of ByteBuffer to aggregate
+     * @return a mono emitting aggregated ByteString
+     */
+    private static Mono<okio.ByteString> aggregate(Flux<ByteBuffer> bbFlux) {
+        Objects.requireNonNull(bbFlux);
+        return Mono.using(okio.Buffer::new,
+                buffer -> bbFlux.reduce(buffer, (b, byteBuffer) -> {
+                    try {
+                        b.write(byteBuffer);
+                        return b;
+                    } catch (IOException ioe) {
+                        throw Exceptions.propagate(ioe);
+                    }
+                })
+                .map(b -> okio.ByteString.of(b.readByteArray())),
+                okio.Buffer::clear)
+            .switchIfEmpty(EMPTY_BYTE_STRING_MONO);
+    }
+
+    /**
+     * An implementation of azure-core HttpResponse for OkHttp.
+     */
+    private static class OkHttpResponse extends HttpResponse {
+        private final okhttp3.Response inner;
+        private final HttpHeaders headers;
+        private final static int BYTE_BUFFER_CHUNK_SIZE = 1024;
+
+        public OkHttpResponse(okhttp3.Response inner, HttpRequest request) {
+            this.inner = inner;
+            this.headers = fromOkHttpHeaders(this.inner.headers());
+            super.request(request);
+        }
+
+        @Override
+        public int statusCode() {
+            return this.inner.code();
+        }
+
+        @Override
+        public String headerValue(String name) {
+            return this.headers.value(name);
+        }
+
+        @Override
+        public HttpHeaders headers() {
+            return this.headers;
+        }
+
+        @Override
+        public Flux<ByteBuffer> body() {
+            return this.responseBody() != null
+                    ? toFluxByteBuffer(this.responseBody().byteStream())
+                    : Flux.empty();
+        }
+
+        @Override
+        public Mono<byte[]> bodyAsByteArray() {
+            if (this.responseBody() == null) {
+                return Mono.empty();
+            } else {
+                return Mono.using(() -> this.responseBody(),
+                        rb -> {
+                            try {
+                                byte[] content = rb.bytes();
+                                return content.length == 0 ? Mono.empty() : Mono.just(content);
```

Comment:
Same here - empty array vs `Mono.empty()`. I am not sure if there's a difference between a `null` responseBody vs a non-null responsebody with empty byte array.

Should we instead return `Mono.just(content)` where content could potentially be empty?

Reply:
Good question @srnagar :)
 
Q1: – *if there's a difference between a null responseBody ..*:
 
`okhttp3.Response::body()` getter will not return null for server returned responses. It can be null only if we build response manually with null body (e.g. for mocking) or for the cases described here [ref](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-response/body/). Those null cases are not applicable for OkHttp HttpClient impl. I added the null check as a defensive check, those might never hit.
 
Q2: - *using Mono.just(content) [where content=new byte[0]] vs Mono.empty()*
 
If content part is absent in a http wire response then content getters in `okhttp3.ResponseBody` instance uses Java empty semantic. i.e. `okhttp3.ResponseBody::bytes()` getter return `new byte[0]`, `okhttp3.ResponseBody::string()` returns "".
 
Similarly Netty uses empty `ByteBuf` i.e. `ByteBuf` of size 0 if content part is absent.
 
In reactive, the pattern of using `Flux::empty()/Mono::empty()` for absence of value is common.
 
If we look at reactor-netty as an example, it translate an empty `ByteBuf` from Netty to `Flux::empty()`.
 
Our reactor-netty azure-core HttpClient plugin follows the same pattern.
 
Any implementation of azure-core HttpClient is expected to follow this pattern, hence okHttp HttpClient plugin.
 
---
 
It looks like I need to clarify above points in the code to make it readable for others. I've opened a [PR](https://github.com/Azure/azure-sdk-for-java/pull/5222)  same behavior but better readability. It also centralize resource management.
 
Thanks for raising this question.